lib/ruby_indexer/lib/ruby_indexer/prefix_tree.rb

```diff
@@ -0,0 +1,145 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyIndexer
+  # A PrefixTree is a data structure that allows searching for partial strings fast. The tree is similar to a nested
+  # hash structure, where the keys are the characters of the inserted strings.
+  #
+  # ## Example
+  # ```ruby
+  # tree = PrefixTree.new(["foo"])
+  # # Internally, the structure is analogous to this, but using nodes:
+  # # {
+  # #   "f" => {
+  # #     "o" => {
+  # #       "o" => value
+  # #     }
+  # #   }
+  # # }
+  # # When we search it, it finds all possible values based on partial (or complete matches):
+  # tree.search("") # => ["foo"]
+  # tree.search("f") # => ["foo"]
+  # tree.search("fo") # => ["foo"]
+  # tree.search("foo") # => ["foo"]
+  # ```
+  #
+  # A PrefixTree is useful for autocomplete, since we always want to find all alternatives while the developer hasn't
+  # finished typing yet. This PrefixTree implementation allows for string keys and any arbitrary value using the generic
+  # `Value` type.
+  #
+  # See https://en.wikipedia.org/wiki/Trie for more information
+  class PrefixTree
+    extend T::Sig
+    extend T::Generic
+
+    Value = type_member
+
+    sig { void }
+    def initialize
+      @root = T.let(Node.new("", ""), Node[Value])
+    end
+
+    # Search the PrefixTree based on a given `prefix`. If `foo` is an entry in the tree, then searching for `fo` will
+    # return it as a result. The result is always an array of the type of value attribute to the generic `Value` type.
+    # Notice that if the `Value` is an array, this method will return an array of arrays, where each entry is the array
+    # of values for a given match
+    sig { params(prefix: String).returns(T::Array[Value]) }
+    def search(prefix)
+      node = find_node(prefix)
+      return [] unless node
+
+      node.collect
+    end
+
+    # Inserts a `value` using the given `key`
+    sig { params(key: String, value: Value).void }
+    def insert(key, value)
+      node = @root
+
+      key.each_char do |char|
+        node = node.children[char] ||= Node.new(char, value, node)
+      end
+
+      node.leaf = true
+    end
+
+    # Deletes the entry identified by `key` from the tree. Notice that a partial match will still delete all entries
+    # that match it. For example, if the tree contains `foo` and we ask to delete `fo`, then `foo` will be deleted
+    sig { params(key: String).void }
+    def delete(key)
```

Comment:
Two entries because the keys are different, even though the value is the same. And it would delete only the one for the file that was deleted.

Reply:
In that case, yes, but that's not the structure we'll use. We'd have this structure instead

```ruby
tree = PrefixTree[T::Array[Entry]].new
tree.insert("Foo::Bar", [#<Entry >, #<Entry >])
```

That's why I allowed overriding values in #955. If one entry is removed for `Foo::Bar`, we just override the value modifying the array.

Reply:
Doesn't that move all the complexity of deleting vs. updating to the client as you do here:

```rb
        if entries.empty?
          @entries.delete(name)
          @entries_tree.delete(name)
        else
          @entries_tree.insert(name, entries)
        end
```

I wonder if instead the PrefixTree could be use `T::Array[Node[Value]]` instead of `Node[T::Array[Value]]`, so we move back the complexity of deletion on the lib rather than the client?

Reply:
Yes exactly, but in exchange you hide the complexity away.

I let you chose 👍

Reply:
Let's keep the behaviour similar to a hash for now. We can always re-evaluate if this becomes a problem.